File: python/src/wordcount.py
"""Count ASCII words in a file."""

import sys
from dataclasses import dataclass
from pathlib import Path
from time import perf_counter_ns

CHECKSUM_OFFSET = 2_166_136_261
CHECKSUM_PRIME = 16_777_619
CHECKSUM_MASK = 0xFFFF_FFFF
USAGE = "usage: wordcount.py [--json] [--top N] [--max-word N] <file>"

type Entry = tuple[bytes, int]
type Result = tuple[int, int, list[Entry]]


@dataclass(frozen=True, slots=True)
class Options:
    path: Path
    top: int = 10
    max_word: int = 1024
    bench_runs: int = 0
    bench_warmups: int = 0
    json: bool = False


def count_words(data: bytes, limit: int, requested_max_word: int) -> Result:
    """Count and rank ASCII words in data."""
    max_word = normalize_max_word(requested_max_word)
    counts: dict[bytes, int] = {}
    total = 0
    cursor = 0
    data_length = len(data)

    while cursor < data_length:
        while cursor < data_length and not 97 <= (data[cursor] | 0x20) <= 122:
            cursor += 1
        start = cursor
        while cursor < data_length and 97 <= (data[cursor] | 0x20) <= 122:
            cursor += 1
        if start == cursor:
            continue

        word = data[start : min(cursor, start + max_word)].lower()
        counts[word] = counts.get(word, 0) + 1
        total += 1

    entries = sorted(counts.items(), key=lambda item: (-item[1], item[0]))
    return total, len(entries), entries[:limit]


def normalize_max_word(value: int) -> int:
    if value == 0:
        return 64
    return min(1024, max(4, value))


def checksum(result: Result) -> int:
    total, unique, entries = result
    value = mix_integer(CHECKSUM_OFFSET, total, 8)
    value = mix_integer(value, unique, 8)
    for word, count in entries:
        for byte in word:
            value = mix_byte(value, byte)
        value = mix_integer(value, count, 8)
    return value


def mix_byte(value: int, byte: int) -> int:
    return ((value ^ byte) * CHECKSUM_PRIME) & CHECKSUM_MASK


def mix_integer(checksum_value: int, value: int, width: int) -> int:
    for _ in range(width):
        checksum_value = mix_byte(checksum_value, value & 0xFF)
        value >>= 8
    return checksum_value


def bench(data: bytes, options: Options) -> str:
    for _ in range(options.bench_warmups):
        checksum(count_words(data, options.top, options.max_word))

    value = CHECKSUM_OFFSET
    started = perf_counter_ns()
    for _ in range(options.bench_runs):
        value = mix_integer(
            value,
            checksum(count_words(data, options.top, options.max_word)),
            4,
        )
    mean_ms = (perf_counter_ns() - started) / options.bench_runs / 1_000_000
    return f'{{"mean_ms":{mean_ms:.6f},"checksum":{value}}}'


def render_json(result: Result) -> str:
    total, unique, entries = result
    top = ",".join(
        f'{{"word":"{word.decode("ascii")}","count":{count}}}' for word, count in entries
    )
    return f'{{"total":{total},"unique":{unique},"top":[{top}]}}'


def render_text(result: Result) -> str:
    total, unique, entries = result
    lines = ["count word", *(f"{count} {word.decode('ascii')}" for word, count in entries)]
    lines.extend((f"total {total}", f"unique {unique}"))
    return "\n".join(lines)


def parse_args(args: list[str]) -> Options:
    path: Path | None = None
    top = 10
    max_word = 1024
    bench_runs = 0
    bench_warmups = 0
    json_output = False
    index = 0

    while index < len(args):
        arg = args[index]
        index += 1
        if arg == "--json":
            json_output = True
        elif arg in {"--top", "--max-word", "--bench-runs", "--bench-warmups"}:
            if index == len(args):
                raise ValueError(USAGE)
            value = parse_number(args[index], arg)
            index += 1
            if arg == "--top":
                top = value
            elif arg == "--max-word":
                max_word = value
            elif arg == "--bench-runs":
                bench_runs = value
            else:
                bench_warmups = value
        elif arg.startswith("--top="):
            top = parse_number(arg.removeprefix("--top="), "--top")
        elif arg.startswith("--max-word="):
            max_word = parse_number(arg.removeprefix("--max-word="), "--max-word")
        elif arg.startswith("--bench-runs="):
            bench_runs = parse_number(arg.removeprefix("--bench-runs="), "--bench-runs")
        elif arg.startswith("--bench-warmups="):
            bench_warmups = parse_number(arg.removeprefix("--bench-warmups="), "--bench-warmups")
        elif arg.startswith("-") or path is not None:
            raise ValueError(USAGE)
        else:
            path = Path(arg)

    if path is None or top == 0:
        raise ValueError(USAGE)
    return Options(path, top, max_word, bench_runs, bench_warmups, json_output)


def parse_number(value: str, name: str) -> int:
    if not value or not value.isascii() or not value.isdecimal():
        raise ValueError(f"wordcount.py: {name} must be a number")
    return int(value)


def main() -> int:
    try:
        options = parse_args(sys.argv[1:])
        data = options.path.read_bytes()
    except (OSError, ValueError) as error:
        print(error, file=sys.stderr)
        return 1

    if options.bench_runs > 0:
        print(bench(data, options))
    else:
        result = count_words(data, options.top, options.max_word)
        print(render_json(result) if options.json else render_text(result))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/src/wordcount.ts
import { FileSystem } from '@effect/platform';
import { NodeContext, NodeRuntime } from '@effect/platform-node';
import { Console, Data, Effect, Schema } from 'effect';

type Entry = Readonly<{ word: string; count: number }>;
type Result = Readonly<{ total: number; unique: number; top: readonly Entry[] }>;
type Options = Readonly<{
  path: string;
  top: number;
  maxWord: number;
  benchRuns: number;
  benchWarmups: number;
  json: boolean;
}>;

const checksumOffset = 2_166_136_261;
const checksumPrime = 16_777_619;
const usage = 'usage: wordcount.ts [--json] [--top N] [--max-word N] <file>';
const resultJson = Schema.parseJson(
  Schema.Struct({
    total: Schema.Number,
    unique: Schema.Number,
    top: Schema.Array(Schema.Struct({ word: Schema.String, count: Schema.Number })),
  }),
);
const benchJson = Schema.parseJson(Schema.Struct({ mean_ms: Schema.Number, checksum: Schema.Number }));

class UsageError extends Data.TaggedError('UsageError')<{ message: string }> {}

function normalizeMaxWord(value: number): number {
  return value === 0 ? 64 : Math.min(1024, Math.max(4, value));
}

function countWords(bytes: Uint8Array, limit: number, requestedMaxWord: number): Result {
  const counts = new Map<string, number>();
  const maxWord = normalizeMaxWord(requestedMaxWord);
  let word = '';
  let total = 0;

  for (const byte of bytes) {
    const lower = byte | 0x20;
    if (lower >= 97 && lower <= 122) {
      if (word.length < maxWord) {
        word += String.fromCharCode(lower);
      }
    } else if (word.length > 0) {
      counts.set(word, (counts.get(word) ?? 0) + 1);
      total += 1;
      word = '';
    }
  }
  if (word.length > 0) {
    counts.set(word, (counts.get(word) ?? 0) + 1);
    total += 1;
  }

  const top = [...counts]
    .map(([entryWord, count]) => ({ word: entryWord, count }))
    .sort((left, right) => {
      const countOrder = right.count - left.count;
      return countOrder !== 0 ? countOrder : left.word < right.word ? -1 : left.word > right.word ? 1 : 0;
    })
    .slice(0, limit);
  return { total, unique: counts.size, top };
}

function mixByte(checksumValue: number, byte: number): number {
  return Math.imul(checksumValue ^ byte, checksumPrime) >>> 0;
}

function mixInteger(checksumValue: number, value: number, width: number): number {
  let mixed = checksumValue;
  let remaining = BigInt(value);
  for (let index = 0; index < width; index += 1) {
    mixed = mixByte(mixed, Number(remaining & 0xffn));
    remaining >>= 8n;
  }
  return mixed;
}

function checksum(result: Result): number {
  let value = mixInteger(checksumOffset, result.total, 8);
  value = mixInteger(value, result.unique, 8);
  for (const entry of result.top) {
    for (let index = 0; index < entry.word.length; index += 1) {
      value = mixByte(value, entry.word.charCodeAt(index));
    }
    value = mixInteger(value, entry.count, 8);
  }
  return value;
}

function renderBench(bytes: Uint8Array, options: Options): string {
  for (let index = 0; index < options.benchWarmups; index += 1) {
    checksum(countWords(bytes, options.top, options.maxWord));
  }

  let value = checksumOffset;
  const started = Bun.nanoseconds();
  for (let index = 0; index < options.benchRuns; index += 1) {
    value = mixInteger(value, checksum(countWords(bytes, options.top, options.maxWord)), 4);
  }
  const meanMs = (Bun.nanoseconds() - started) / 1_000_000 / options.benchRuns;
  return Schema.encodeSync(benchJson)({ mean_ms: meanMs, checksum: value });
}

function renderText(result: Result): string {
  const lines = ['count word'];
  for (const entry of result.top) {
    lines.push(`${entry.count} ${entry.word}`);
  }
  lines.push(`total ${result.total}`, `unique ${result.unique}`);
  return lines.join('\n');
}

function parseNumber(value: string | undefined, name: string): number {
  if (value === undefined || !/^\d+$/.test(value)) {
    throw new Error(`${name} must be a number`);
  }
  const parsed = Number(value);
  if (!Number.isSafeInteger(parsed)) {
    throw new Error(`${name} must be a number`);
  }
  return parsed;
}

function parseArgs(args: readonly string[]): Options {
  let path: string | undefined;
  let top = 10;
  let maxWord = 1024;
  let benchRuns = 0;
  let benchWarmups = 0;
  let json = false;

  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (arg === undefined) {
      throw new Error(usage);
    }
    if (arg === '--json') {
      json = true;
    } else if (arg === '--top' || arg === '--max-word' || arg === '--bench-runs' || arg === '--bench-warmups') {
      const value = parseNumber(args[++index], arg);
      if (arg === '--top') {
        top = value;
      } else if (arg === '--max-word') {
        maxWord = value;
      } else if (arg === '--bench-runs') {
        benchRuns = value;
      } else {
        benchWarmups = value;
      }
    } else if (arg.startsWith('--top=')) {
      top = parseNumber(arg.slice(6), '--top');
    } else if (arg.startsWith('--max-word=')) {
      maxWord = parseNumber(arg.slice(11), '--max-word');
    } else if (arg.startsWith('--bench-runs=')) {
      benchRuns = parseNumber(arg.slice(13), '--bench-runs');
    } else if (arg.startsWith('--bench-warmups=')) {
      benchWarmups = parseNumber(arg.slice(16), '--bench-warmups');
    } else if (arg.startsWith('-') || path !== undefined) {
      throw new Error(usage);
    } else {
      path = arg;
    }
  }

  if (path === undefined || top === 0) {
    throw new Error(usage);
  }
  return { path, top, maxWord, benchRuns, benchWarmups, json };
}

const program = Effect.gen(function* () {
  const options = yield* Effect.try({
    try: () => parseArgs(process.argv.slice(2)),
    catch: (cause) =>
      new UsageError({
        message: cause instanceof Error ? cause.message : String(cause),
      }),
  });
  const fileSystem = yield* FileSystem.FileSystem;
  const bytes = yield* fileSystem.readFile(options.path);
  if (options.benchRuns > 0) {
    yield* Console.log(renderBench(bytes, options));
  } else {
    const result = countWords(bytes, options.top, options.maxWord);
    const output = options.json ? yield* Schema.encode(resultJson)(result) : renderText(result);
    yield* Console.log(output);
  }
});

NodeRuntime.runMain(
  program.pipe(
    Effect.tapError((error) => Console.error(error.message)),
    Effect.provide(NodeContext.layer),
  ),
  { disableErrorReporting: true },
);
